File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.ts
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    
    // BRUTE FORCE SOLUTION O(n^2)

    // productExceptSelf(nums: number[]): number[] {
    //     var output = [];
    //     for (var i = 0; i < nums.length; i++) {
    //         var product = 1;
    //         for (var j = 0; j < nums.length; j++) {
    //             if (i != j){
    //                 product *= nums[j]
    //             }
    //         }
    //         output.push(product)
    //     }
    //     return output;
    // }

    // compute prefix product and suffix product at each index 
    // then loop over nums and multiple prefix and suffix products to get result
    productExceptSelf(nums: number[]): number[] {
        var output = [];
        var prefixProd = {};
        var suffixProd = {};

        var product = 1;
        for (var i = 0; i < nums.length; i++) {
            if (i != 0){
                product *= nums[i-1];
            }
            prefixProd[i] = product;
        }

        product = 1;
        for (var i = nums.length -1; i >= 0; i--) {
            if (i != nums.length-1){
                product *= nums[i+1];
            }
            suffixProd[i] = product;
        }

        product = 1;
        for (var j = 0; j<nums.length; j++) {
            output.push(prefixProd[j] * suffixProd[j]);
        }
        return output;
    }
}

File: Data Structures & Algorithms/subsets/submission-1.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> output = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        generate(output, curr, 0, nums);
        return output;

    }

    private void generate(List<List<Integer>> output, List<Integer> curr, int index, int[] nums) {
        if(index >= nums.length){
            output.add(new ArrayList<>(curr));
            return;
        }
        curr.add(nums[index]);
        generate(output, curr, index+1, nums);

        curr.removeLast();
        generate(output, curr, index+1, nums);
    }
}
